rust/main/agents/relayer/src/msg/metadata/message_builder.rs

```diff
@@ -0,0 +1,427 @@
+#![allow(clippy::blocks_in_conditions)] // TODO: `rustc` 1.80.1 clippy issue
+#![allow(clippy::unnecessary_get_then_check)] // TODO: `rustc` 1.80.1 clippy issue
+
+use std::{fmt::Debug, sync::Arc};
+
+use async_trait::async_trait;
+use eyre::{Context, Result};
+use hyperlane_core::{HyperlaneMessage, InterchainSecurityModule, ModuleType, H256};
+
+use tokio::sync::Mutex;
+use tracing::instrument;
+
+use crate::msg::metadata::base_builder::BaseMetadataBuilderTrait;
+
+use super::{
+    aggregation::AggregationIsmMetadataBuilder,
+    base::{IsmWithMetadataAndType, MetadataBuildError},
+    ccip_read::CcipReadIsmMetadataBuilder,
+    metadata_builder::MessageMetadataBuildParams,
+    multisig::{MerkleRootMultisigMetadataBuilder, MessageIdMultisigMetadataBuilder},
+    routing::RoutingIsmMetadataBuilder,
+    Metadata, MetadataBuilder,
+};
+
+/// Builds metadata for a message.
+#[derive(Debug, Clone)]
+pub struct MessageMetadataBuilder {
+    pub base: Arc<Box<dyn BaseMetadataBuilderTrait>>,
+    pub app_context: Option<String>,
+    /// current ISM depth.
+    /// ISMs can be structured recursively. We keep track of the depth
+    /// of the recursion to avoid infinite loops.
+    pub depth: u32,
+    /// current ISM count
+    pub ism_count: Arc<Mutex<u32>>,
+}
+
+#[async_trait]
+impl MetadataBuilder for MessageMetadataBuilder {
+    #[instrument(err, skip(self, message), fields(destination_domain=self.base_builder().destination_domain().name()))]
+    async fn build(
+        &self,
+        message: &HyperlaneMessage,
+        params: MessageMetadataBuildParams,
+    ) -> eyre::Result<Metadata> {
+        build_message_metadata(self, message, params)
+            .await
+            .map(|res| res.metadata)
+    }
+}
+
+impl MessageMetadataBuilder {
+    pub async fn new(
+        base: Arc<Box<dyn BaseMetadataBuilderTrait>>,
+        ism_address: H256,
+        message: &HyperlaneMessage,
+    ) -> Result<Self> {
+        let app_context = base
+            .app_context_classifier()
+            .get_app_context(message, ism_address)
+            .await?;
+        Ok(Self {
+            base,
+            app_context,
+            depth: 0,
+            ism_count: Arc::new(Mutex::new(0)),
+        })
+    }
+
+    pub fn base_builder(&self) -> &Arc<Box<dyn BaseMetadataBuilderTrait>> {
+        &self.base
+    }
+}
+
+/// Builds metadata for a message.
+pub async fn build_message_metadata(
+    message_builder: &MessageMetadataBuilder,
+    message: &HyperlaneMessage,
+    params: MessageMetadataBuildParams,
+) -> eyre::Result<IsmWithMetadataAndType> {
+    let res = build_message_metadata_recursive(message_builder.clone(), message, params).await?;
+    Ok(res)
```

Comment:
I wonder if this will work here:

```suggestion
    build_message_metadata_recursive(message_builder.clone(), message, params).await
```